walkeeper/src/safekeeper.rs

```diff
@@ -825,6 +829,45 @@ where
             self.append_response(),
         )))
     }
+
+    /// Update timeline state with peer safekeeper data.
+    pub fn record_safekeeper_info(&mut self, sk_info: &SafekeeperInfo) -> Result<()> {
+        let mut sync_control_file = false;
+        if let (Some(commit_lsn), Some(last_log_term)) = (sk_info.commit_lsn, sk_info.last_log_term)
+        {
+            // Note: the check is too restrictive, generally we can update local
+            // commit_lsn if our history matches (is part of) history of advanced
+            // commit_lsn provider.
+            if last_log_term == self.get_epoch() {
+                self.global_commit_lsn = max(commit_lsn, self.global_commit_lsn);
+                self.update_commit_lsn()?;
+            }
+        }
+        if let Some(s3_wal_lsn) = sk_info.s3_wal_lsn {
+            let new_s3_wal_lsn = max(s3_wal_lsn, self.inmem.s3_wal_lsn);
+            sync_control_file |=
```

Comment:
This check can always be false, because it checks only in-memory delta which is updated right after. The same goes for the next wal_seg_size check.